quickwit/quickwit-directories/src/caching_directory.rs

```diff
@@ -30,40 +30,44 @@ use tantivy::{Directory, HasLen};
 
 /// The caching directory is a simple cache that wraps another directory.
 #[derive(Clone)]
-pub struct CachingDirectory {
+pub struct CachingDirectory<C> {
     underlying: Arc<dyn Directory>,
-    // TODO fixme: that's a pretty ugly cache we have here.
-    cache: Arc<ByteRangeCache>,
+    cache: C,
 }
 
-impl CachingDirectory {
-    /// Creates a new CachingDirectory.
+impl CachingDirectory<Arc<ByteRangeCache>> {
+    /// Creates a new CachingDirectory with a default cache.
     ///
     /// Warming: The resulting CacheDirectory will cache all information without ever
     /// removing any item from the cache.
-    pub fn new_unbounded(underlying: Arc<dyn Directory>) -> CachingDirectory {
-        CachingDirectory {
-            underlying,
-            cache: Arc::new(ByteRangeCache::with_infinite_capacity(
-                &quickwit_storage::STORAGE_METRICS.shortlived_cache,
-            )),
-        }
+    pub fn new_unbounded(underlying: Arc<dyn Directory>) -> CachingDirectory<Arc<ByteRangeCache>> {
+        let byte_range_cache = ByteRangeCache::with_infinite_capacity(
+            &quickwit_storage::STORAGE_METRICS.shortlived_cache,
+        );
+        CachingDirectory::new(underlying, Arc::new(byte_range_cache))
+    }
+}
+
+impl<C: DirectoryCache> CachingDirectory<C> {
+    /// Creates a new CachingDirectory with an existing cache.
+    pub fn new(underlying: Arc<dyn Directory>, cache: C) -> CachingDirectory<C> {
+        CachingDirectory { underlying, cache }
     }
 }
 
-impl fmt::Debug for CachingDirectory {
+impl<T> fmt::Debug for CachingDirectory<T> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "CachingDirectory({:?})", self.underlying)
     }
 }
 
-struct CachingFileHandle {
+struct CachingFileHandle<C> {
```

Comment:
what is the extra complexity for?

Reply:
To avoid dropping the permits early or late compared to the actual memory dealocation, I want to attach the `SearchPermit`'s lifecycle to the `ByteRangeCache`. But this wasn't easy because both `ByteRangeCache` and the abstraction above `CachingDirectory` are used in multiple place, with and without a permit. Making the `CachingDirectory` support both a cache instance that is tied to a permit or not is the simplest way I found. Then remained the question of dynamic VS static dispatch. I opted for static dispatch because I found that its explicit display of the kind of cache used on the caller side is actually useful. It makes it clear whether we use a cache that holds a permit or not, and exhibits well which allocations are controlled by the memory pool and which aren't.

The specific type parameter on `CachingFileHandle<C>` that you highlight here is a collateral damage (required because file handles contain a reference to the directory), but it's pretty minor given that it is a private data structure.

Reply:
I don't think this is a good addition. I understand the point of tying search permit to the byte range, but:
- this is a bit overkill
- it gives the misleading impression that the memory permit is reflecting accurately the content of the byte range cache. This is not the case.

Reply:
Very good point. I'll remove it.